runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10412,227 +10412,732 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+                  generateTrg1Src1Instruction(cg, TR::InstOpCode::extsb, node, tempReg, tempReg);
+                  }
+               break;
+            case TR::Int16:
+               generateTrg1MemInstruction(cg, isSigned ? TR::InstOpCode::lha : TR::InstOpCode::lhz, node, tempReg,
+                     TR::MemoryReference::createWithDisplacement(cg, valueReg, i*2, 2));
+               break;
+            default:
+               TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+            }
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, hashReg, hashReg, tempReg);
+         }
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, valueReg, valueReg, elementSize * unrollFactor);
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, serialUnrollLabel);
+
+      // --- single loop
+      generateLabelInstruction(cg, TR::InstOpCode::label, node, serialLoopLabel);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp8, node, condReg, valueReg, endReg);
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bge, node, endLabel, condReg);
+      generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, tempReg, hashReg, 5, 0xFFFFFFFFFFFFFFE0);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, hashReg, hashReg, tempReg);
+      switch (elementType)
+         {
+         case TR::Int8:
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lbzx, node, tempReg,
+               TR::MemoryReference::createWithIndexReg(cg, NULL, valueReg, 1));
+            if (isSigned)
+               {
+               generateTrg1Src1Instruction(cg, TR::InstOpCode::extsb, node, tempReg, tempReg);
+               }
+            break;
+         case TR::Int16:
+            generateTrg1MemInstruction(cg, isSigned ? TR::InstOpCode::lhax : TR::InstOpCode::lhzx, node, tempReg,
+               TR::MemoryReference::createWithIndexReg(cg, NULL, valueReg, 2));
+            break;
+         case TR::Int32:
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lwzx, node, tempReg,
+               TR::MemoryReference::createWithIndexReg(cg, NULL, valueReg, 4));
+            break;
+         default:
+            TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+         }
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, hashReg, hashReg, tempReg);
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, valueReg, valueReg, elementSize);
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, serialLoopLabel);
+      }
+
+   // special cases
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, special2Label);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, vendReg, 2*elementSize);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, special3Label, condReg);
+   // length 2
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::mulli, node, hashReg, hashReg, 961);
```

Comment:
use the immediate offset load instructions (such as lbz, lhz/lha, and lwz) for better performance and you don't need to keep updating valueReg which can be updated less frequently.